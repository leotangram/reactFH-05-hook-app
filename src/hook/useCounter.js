import { useState } from 'react'

const useCounter = (initialCounter = 10) => {
  const [counter, setCounter] = useState(initialCounter)

  const increment = (factor = 1) => setCounter(counter + factor)
  const decrement = (factor = 1) => setCounter(counter - factor)
  const reset = () => setCounter(initialCounter)

  return { counter, increment, decrement, reset }
}

export default useCounter
