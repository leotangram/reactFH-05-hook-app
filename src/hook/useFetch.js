import { useState, useEffect, useRef } from 'react'

export const useFetch = url => {
  const isMountedRef = useRef(true)

  useEffect(() => {
    return () => {
      isMountedRef.current = false
    }
  }, [])

  const [data, setData] = useState({ data: null, loading: true, error: null })
  useEffect(() => {
    setData({ data: null, loading: true, error: null })
    fetch(url)
      .then(response => response.json())
      .then(data => {
        isMountedRef.current && setData({ loading: false, error: null, data })
      })
  }, [url])

  return data
}
