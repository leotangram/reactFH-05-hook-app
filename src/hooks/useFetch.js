import { useState, useEffect, useRef } from 'react'

export const useFetch = url => {
  const isMountedRef = useRef(true)

  useEffect(() => {
    return () => {
      isMountedRef.current = false
    }
  }, [])

  const [data, setData] = useState({ data: null, loading: true, error: null })
  useEffect(() => {
    setData({ data: null, loading: true, error: null })
    fetch(url)
      .then(response => response.json())
      .then(data => {
        isMountedRef.current && setData({ loading: false, error: null, data })
      })
      .catch(() => {
        setData({
          data: null,
          loading: false,
          error: 'No se pudo cargar la info'
        })
      })
  }, [url])

  return data
}
