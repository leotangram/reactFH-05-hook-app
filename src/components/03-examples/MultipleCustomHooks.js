import React from 'react'
import '../02-useEffect/effects.css'
import { useFetch } from '../../hook/useFetch'

const MultipleCustomHooks = () => {
  const data = useFetch(`https://www.breakingbadapi.com/api/quotes/1`)
  console.log(data)

  return (
    <div>
      <h2>CustomHooks</h2>
    </div>
  )
}

export default MultipleCustomHooks
