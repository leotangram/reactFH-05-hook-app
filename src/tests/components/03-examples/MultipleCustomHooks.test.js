import React from 'react'
import { shallow } from 'enzyme'
import { useFetch } from '../../../hooks/useFetch'
import { useCounter } from '../../../hooks/useCounter'
import MultipleCustomHooks from '../../../components/03-examples/MultipleCustomHooks'

jest.mock('../../../hooks/useFetch')
jest.mock('../../../hooks/useCounter')

describe('<MultipleCustomHooks />', () => {
  useCounter.mockReturnValue({
    counter: 10,
    increment: () => {}
  })

  test('should show correctly', () => {
    useFetch.mockReturnValue({
      data: null,
      loading: true,
      error: null
    })
    const wrapper = shallow(<MultipleCustomHooks />)
    expect(wrapper).toMatchSnapshot()
  })

  test('should show the info', () => {
    useFetch.mockReturnValue({
      data: [
        {
          author: 'Leonardo',
          quote: 'Holi'
        }
      ],
      loading: false,
      error: null
    })

    const wrapper = shallow(<MultipleCustomHooks />)

    expect(wrapper.find('.alert').exists()).toBe(false)
    expect(wrapper.find('.mb-0').text().trim()).toBe('Holi')
    expect(wrapper.find('footer').text().trim()).toBe('Leonardo')

    console.log('wrapper: ', wrapper.html())
  })
})
