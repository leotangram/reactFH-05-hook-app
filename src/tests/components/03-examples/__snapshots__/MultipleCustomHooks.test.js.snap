// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MultipleCustomHooks /> should show correctly 1`] = `
<div>
  <h2>
    CustomHooks
  </h2>
  <h3>
    BrakingBad Quotes
  </h3>
  <hr />
  <div
    className="alert alert-info text-center"
  >
    loading...
  </div>
  <button
    className="btn btn-primary"
    onClick={[Function]}
  >
    Siguiente frase
  </button>
</div>
`;
